returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,253 @@
+        :param map_seq: post processor function operating on the single-segment level.
+        :param map_seq_stream: post processor function operating on the multiple segment level via an iterator.
+            Allows merging multiple segments into one, or generating multiple output segments from one input segment.
+            When both functions are specified, the segment-level function is applied first to every segment, and the
+            results are passed to the iterator-level function.
+        :param map_outputs: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        assert (
+            self.seq_ordering == "default"
+        ), f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}"
+        assert (
+            map_seq is not None or map_seq_stream is not None
+        ), "need to either define map_seq or map_seq_stream functions"
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq
+        self._map_seq_stream = map_seq_stream
+        self._map_outputs = map_outputs
+
+        self._dataset: Optional[Dataset] = None
+        self._data_keys: Optional[List[str]] = None
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._default_input: Optional[str] = None
+        self._dim_cache: Dict[str, Tuple[List[Dim], Optional[Dim]]] = {}
+
+        if _meta_info_cache:
+            self._default_input = _meta_info_cache["_default_input"]
+            self._dim_cache = _meta_info_cache["_dim_cache"]
+            self._estimated_num_seqs = _meta_info_cache["_estimated_num_seqs"]
+            self.labels = _meta_info_cache["labels"]
+            self.num_inputs = _meta_info_cache["num_inputs"]
+            self.num_outputs = _meta_info_cache["num_outputs"]
+
+    def initialize(self):
+        """init"""
+        self._lazy_init_dataset()
+        super().initialize()
+
+    @property
+    def _meta_info_cache(self) -> Optional[Dict[str, Any]]:
+        if self.num_outputs is None:
+            return None
+        return {
+            "_default_input": self._default_input,
+            "_dim_cache": self._dim_cache,
+            "_estimated_num_seqs": self._estimated_num_seqs,
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
+
+    def _lazy_init_dataset(self):
+        if self._dataset is not None:
+            return
+
+        self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self.labels = self._dataset.labels
```

Comment:
That's also wrong in general. You should read it from `map_outputs`.

Reply:
How?

Reply:
Assuming you have `self._map_outputs_tensor_dict` or so:
```suggestion
        self.labels = {}
        for k, v in self._map_outputs_tensor_dict.data.items():
            if v.vocab:
                self.labels[k] = v.vocab.labels
            elif v.sparse_dim:  # sparse_dim but not vocab
                self.labels[k] = list(map(str, range(v.sparse_dim.dimension)))  # dummy labels
```